### ISSUE TITLE
Value.toSortedSet(Comparator) makes data lost

### ISSUE THREAD

**Issue by mincong-h** [vavr-io/vavr#2413]
`Value#toSortedSet(Comparator)` makes data lost when comparator is **not** correctly implemented. Here're three samples illustrating the problem, written with HashMap in VAVR and Map in Java 11.

In VAVR 0.9.2:

```java
HashMap.of("k1", 1, "k2", 1)
       .toSortedSet((t, u) -> t._2 - u._2);
// TreeSet((k2, 1))
```

```java
HashMap.of("k1", 1, "k2", 1)
       .toSortedSet((t, u) -> t._2 - u._2 != 0 ? t._2 - u._2 : t._1.compareTo(u._1));
// TreeSet((k1, 1), (k2, 1))
```

In Java 11:

```java
Map.of("k1", 1, "k2", 1).entrySet().stream()
   .sorted((t, u) -> t.getValue() - u.getValue())
   .collect(Collectors.toSet());
// [k2=1, k1=1]
```

I think the problem might come from comparison logic: two values of type T are considered as the same when the comparator returns 0.

**Comment by danieldietrich** [comment]
## TL;DR

Java 11 HashMap has no notion of (key) order. Your sorted() call does effectively nothing. You just produce a HashSet of entries.

In your VAVR create a SortedSet by deduplicating the entries using (a custom comparator).

---

## Detailed

You do two different things. Let me recap:

**Your Java 11 code:**

| Step | Action | Result |
| --- | --- | --- |
| 1 | Map.of("k1", 1, "k2", 1) | {k2=1, k1=1} |
| 2 | .entrySet() | Set of entries [k2=1, k1=1] |
| 3 | .stream() | lazy ops pipeline |
| 4 | .sorted((t, u) -> t.getValue() - u.getValue()) | Sort by entry values, duplicates allowed. Effectively does nothing because all values are equal. |
| 5 | .collect(Collectors.toSet()) | **HashSet** of entries [k2=1, k1=1] |

In step 4, you change the order of elements by comparing their values. **Especially, duplicates are allowed here.** It is only the order of elements that might be changed. However, in this example the entry values are equal, so sorted() does effectively nothing.

Regarding the [Java 11 API docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.Entry.html#equals(java.lang.Object)), Java Map.Entry#equals is defined to be

```java
(e1.getKey()==null ?
      e2.getKey()==null : e1.getKey().equals(e2.getKey()))  &&
     (e1.getValue()==null ?
      e2.getValue()==null : e1.getValue().equals(e2.getValue()))
```

Namely, in Java 11 map entries are equal, if they are component wise equal. In step 5 you create a **Java HashSet**.  Elements are only inserted during the build process, if there does not exist an already inserted element that is equal resp. the equals() method. Because of component-wise comparison, all elements in your example are different. Therefore you see all entries of the original HashMap.

**Your Vavr 11 code:**

| Step | Action | Result |
| --- | --- | --- |
| 1 | HashMap.of("k1", 1, "k2", 1) | HashMap((k1, 1), (k2, 1)) |
| 2 | .toSortedSet((t, u) -> t._2 - u._2) | **TreeSet** of tuples TreeSet((k2, 1)) |

In step 1 you create a HashMap of tuples.

In step 2 you create a Vavr TreeSet of tuples that only inserts a tuple, if there does not exist an already inserted tuple that is equal regarding the given Comparator. **Especially, NO duplicates are allowed here.**

Because the Comparator considers only the tuple values and the values are all the same, you see only one resulting tuple in the TreeSet.

In your second example, `.toSortedSet((t, u) -> t._2 - u._2 != 0 ? t._2 - u._2 : t._1.compareTo(u._1))` you effectively create a Comparator, that is roughly equivalent to the definition of Java's Map.Entry#equals semantics (see above) _for the equal case_.

---

## Bottomline

Nothing gets lost. All works as specified by the API spec of the VAVR collections. Your can't compare the solutions you gave. They have a different purpose. VAVR produces a TreeSet, Java produces a HashSet.
